src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -93,6 +82,41 @@ private static bool IsValueWriteAtomic()
             }
         }
 
+        /// <summary>Determines whether type TValue can be written atomically.</summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsValueWriteAtomic()
+        {
+            // Section 12.6.6 of ECMA CLI explains which types can be read and written atomically without
+            // the risk of tearing. See https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-335.pdf
+
+            // Check common cases that JIT can intrinsify even in shared generic code up front.
+            if (!typeof(TValue).IsValueType ||
+                typeof(TValue) == typeof(IntPtr) ||
+                typeof(TValue) == typeof(UIntPtr) ||
+                typeof(TValue) == typeof(bool) ||
+                typeof(TValue) == typeof(sbyte) ||
+                typeof(TValue) == typeof(short) ||
+                typeof(TValue) == typeof(int) ||
+                typeof(TValue) == typeof(byte) ||
+                typeof(TValue) == typeof(ushort) ||
+                typeof(TValue) == typeof(uint) ||
+                typeof(TValue) == typeof(char) ||
+                typeof(TValue) == typeof(float))
+            {
+                return true;
+            }
+
+            if (IntPtr.Size == 8 &&
+                (typeof(TValue) == typeof(long) ||
+                 typeof(TValue) == typeof(ulong) ||
+                 typeof(TValue) == typeof(double)))
+            {
+                return true;
+            }
+
+            return s_isValueUnderlyingTypeWriteAtomic;
```

Comment:
Curious, why is this fallback field still needed? Couldn't it have simply been `false`?

Reply:
It's needed to avoid regressing enums -- they also return true via this check.

I can get a few more cases here by extracting the field into a separate generic class I think.

Reply:
Adding a comment explaining why that fallback is necessary would help.

Reply:
I've essentially kept the existing code now except extracted it into a new generic class only keyed on the value type. Then I have inlined the shared generic case since there are some limitations around inlining with different shared generics.

Can you take another look? Do you want me to add comment that the check also works for enums?

Reply:
Looks good. I'm guessing you've reverted to type codes for some of the primitives because of these limitations? 

Could you update your benchmark results just in case?

Reply:
The `typeof` checks were unnecessary once I factored the field out into a new static class on only the value type. We can rely on JIT baking in the value of the static readonly field in tier-1 code as long as `TValue` is a value type, so we only need the reference type check up front after that.

Updated the benchmarks.